core/src/main/java/org/apache/iceberg/rest/OAuth2Util.java

```diff
@@ -0,0 +1,142 @@
+  }
+
+  public static String toScope(List<String> scopes) {
+    return SCOPE_JOINER.join(scopes);
+  }
+
+  public static Map<String, String> tokenExchangeRequest(String subjectToken, String subjectTokenType,
+                                                         List<String> scopes) {
+    return tokenExchangeRequest(subjectToken, subjectTokenType, null, null, scopes);
+  }
+
+  public static Map<String, String> tokenExchangeRequest(String subjectToken, String subjectTokenType,
+                                                         String actorToken, String actorTokenType,
+                                                         List<String> scopes) {
+    Preconditions.checkArgument(VALID_TOKEN_TYPES.contains(subjectTokenType),
+        "Invalid token type: %s", subjectTokenType);
+    Preconditions.checkArgument(actorToken == null || VALID_TOKEN_TYPES.contains(actorTokenType),
+        "Invalid token type: %s", actorTokenType);
+
+    ImmutableMap.Builder<String, String> formData = ImmutableMap.builder();
+    formData.put(GRANT_TYPE, TOKEN_EXCHANGE);
+    formData.put(SCOPE, toScope(scopes));
+    formData.put(SUBJECT_TOKEN, subjectToken);
+    formData.put(SUBJECT_TOKEN_TYPE, subjectTokenType);
+    if (actorToken != null) {
+      formData.put(ACTOR_TOKEN, actorToken);
+      formData.put(ACTOR_TOKEN_TYPE, actorTokenType);
+    }
+
+    return formData.build();
+  }
+
+  private static Pair<String, String> parseCredential(String credential) {
+    Preconditions.checkNotNull(credential, "Invalid credential: null");
+    List<String> parts = CREDENTIAL_SPLITTER.splitToList(credential);
+    switch (parts.size()) {
+      case 2:
+        // client ID and client secret
+        return Pair.of(parts.get(0), parts.get(1));
+      case 1:
+        // client ID
+        return Pair.of(null, parts.get(0));
+      default:
+        // this should never happen because the credential splitter is limited to 2
+        throw new IllegalArgumentException("Invalid credential: " + credential);
+    }
+  }
+
+  public static Map<String, String> clientCredentialsRequest(String credential, List<String> scopes) {
+    Pair<String, String> credentialPair = parseCredential(credential);
+    return clientCredentialsRequest(credentialPair.first(), credentialPair.second(), scopes);
+  }
+
+  public static Map<String, String> clientCredentialsRequest(String clientId, String clientSecret,
+                                                             List<String> scopes) {
+    ImmutableMap.Builder<String, String> formData = ImmutableMap.builder();
+    formData.put(GRANT_TYPE, CLIENT_CREDENTIALS);
+    if (clientId != null) {
+      formData.put(CLIENT_ID, clientId);
+    }
+    formData.put(CLIENT_SECRET, clientSecret);
+
+    ImmutableList.Builder<String> scopeBuilder = ImmutableList.builder();
+    scopeBuilder.addAll(scopes);
+    scopeBuilder.add(CATALOG);
+    formData.put(SCOPE, SCOPE_JOINER.join(scopes));
```

Comment:
Nit: Could call `toScope(scopes)` here to reduce the usage of `SCOPE_JOINER` to _just_ that function.

Reply:
Yeah, this looks suspicious to me. I'll rework it.